### PR TITLE
Initial implementation of test suite for npm registry implementations

### PR DESCRIPTION
**Pull request by bajtos**
The goal of this module is to provide a test suite that checks whether an alternative registry implementation is compatible with the official npmjs registry. By keeping the tests outside of any implementation, we can use it later to test various components of our private registry solution.

To ensure the tests are up to date, `npm test` starts a Couch DB instance, uploads the official npmjs registry app and run the test suite against this registry.

/to @sam-github I suppose you are the best person to review this code. If you think there is no need to review this, then I am fine with that too.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,3 +2,76 @@
 
 Validate your npm registry implementation,
 check if it correctly implements the npm protocol.
+
+## Usage
```
Comment (rmg) [inline README.md:6]: This seems like it would primarily be a CLI tool rather than something that would be used programatically, so I think I'd be strongly inclined to focus on the global/CLI usage scenario in the documentation.

**Comment by rmg** [comment]
TL;DR: :shipit: Looks like a good starting point; my complaints should **not** be considered blocking.

Complaint: The mocha/bluebird example is _far_ too prescriptive in my opinion. If these are actual dependencies of the API, then the module should be renamed to something more like 'mocha-npm-registry-validator-as-promised'. But as I said in another comment, this seems far more useful as a CLI tool than an API.

Comment: For the CouchDB config, if the bind is changed to `0.0.0.0` then the vhosts should be able to use any of the `127.0.0.0/8` addresses to avoid the vhost from masking direct CouchDB access. Note that on Mac, you need to add the specific addresses as aliases on loopback via `sudo ifconfig lo0 alias 127.0.0.2`. I tried to verify this on Windows, but I couldn't even get it to connect via `127.0.0.1`, only `localhost`, so I'm not sure what the story there is. Given the `couchdb` on $PATH requirement I'm not sure Windows is even supported as is though, so I'm not too concerned.

**Comment by rmg** [comment]
@bajtos I've got a branch based off this one ready to turn into a pull request once this PR is merged.

**Comment by bajtos** [comment]
> This seems like it would primarily be a CLI tool rather than something that would be used programatically, so I think I'd be strongly inclined to focus on the global/CLI usage scenario in the documentation.

My initial thoughts was to add these tests to a Mocha test suite for a registry implementation (Reggie, Sinopia, etc.)

On the second thought I agree that may not be the most common usage.

I still think there should be an easy way how to run the tests suite programatically, so that you can write a test script that a) starts your registry b) runs the validation.

> Complaint: The mocha/bluebird example is far too prescriptive in my opinion. If these are actual dependencies of the API, then the module should be renamed to something more like 'mocha-npm-registry-validator-as-promised'. But as I said in another comment, this seems far more useful as a CLI tool than an API.

Agreed, mocha/bluebird should be an implementation detail. I'll submit another PR to fix that.

> Comment: For the CouchDB config, if the bind is changed to `0.0.0.0` then the vhosts should be able to use any of the `127.0.0.0/8` addresses to avoid the vhost from masking direct CouchDB access. Note that on Mac, you need to add the specific addresses as aliases on loopback via `sudo ifconfig lo0 alias 127.0.0.2`. I tried to verify this on Windows, but I couldn't even get it to connect via `127.0.0.1`, only `localhost`, so I'm not sure what the story there is. Given the `couchdb` on $PATH requirement I'm not sure Windows is even supported as is though, so I'm not too concerned.

AFAIK `0.0.0.0` is not supported by Windows. E.g. you can't use `http://0.0.0.0/debug?port=5858` to start Node Inspector, you have to type `127.0.0.1` or `localhost`.

I'd like to keep the number of prerequisites as little as possible. Asking developers to run `sudo ifconfig lo0 alias 127.0.0.2` just to be able to run the self-test is more that I would like.

(Context: at the moment, the vanilla CouchDB interface is at `http://localhost/`, the database API for the registry database at `http://localhost/registry`, the npm REST Api at `http://127.0.0.1/`).

Considering all of the above, it's probably best to get rid of the vhost config and use the full long URL as the npm registry. What do you reckon?

**Comment by rmg** [comment]
Removing vhosts seems reasonable.
